### ISSUE TITLE
Nalu-Wind seg fault on Summit regression tests

### ISSUE THREAD

**Issue by PaulMullowney** [Exawind/nalu-wind#929]
Nalu-Wind reg tests are segfaulting on Summit. The output below is for a RELEASE build. RELWITHDEBINFO builds show other issues. I think we should solve this one first as it happens during mesh initialization.

I'm using gcc 9.3.0 and cuda 11.0.3
   Nalu-Wind SHA: 3343b72905
   Trilinos SHA :      dd53a1e0fbe

In the build/reg_tests/airfoilRANSEdgeNGPTrilinos.rst directory, I generate a valgrind callstack via:

export CUDA_LAUNCH_BLOCKING=1 && jsrun --smpiargs="-gpu" -n 1 -a 1 -c 1 -g 1 valgrind --tool=memcheck --leak-check=full --log-file=trilinos_vg%p ../../../naluX  -I ../../../../reg_tests/test_files/airfoilRANSEdgeNGPTrilinos.rst/airfoilRANSEdgeNGPTrilinos.rst.yaml -o airfoilRANSEdgeNGPTrilinos.rst.log

I've attached the log file and vg output. Here's the relevant call stack (without line numbers).

==3068855== Jump to the invalid address stated on the next line
==3068855==    at 0x0: ???
==3068855==    by 0x1110C093: int sierra::nalu::nodes_per_entity<sierra::nalu::ElemDataRequestsGPU>(sierra::nalu::ElemDataRequestsGPU const&, sierra::nalu::METype) [clone .constprop.0] (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==    by 0x1112BD87: void sierra::nalu::nalu_ngp::run_face_elem_algorithm<stk::mesh::DeviceMesh, sierra::nalu::nalu_ngp::FieldManager, sierra::nalu::ElemDataRequests, __nv_hdl_wrapper_t<false, false, __nv_dl_tag<void (sierra::nalu::WallFuncGeometryAlg<sierra::nalu::AlgTraitsFaceElem<sierra::nalu::AlgTraitsEdge_2D, sierra::nalu::AlgTraitsQuad4_2D> >::*)(), &sierra::nalu::WallFuncGeometryAlg<sierra::nalu::AlgTraitsFaceElem<sierra::nalu::AlgTraitsEdge_2D, sierra::nalu::AlgTr\
aitsQuad4_2D> >::execute, 1u>, void (sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh>&), unsigned int const, unsigned int const, sierra::nalu::MasterElement*, sierra::nalu::MasterElement*, bool, double, sierra::nalu::nalu_ngp::impl::ElemFieldOp<stk::mesh::DeviceMesh, stk::mesh::DeviceField<double, stk::mesh::EmptyNgpFieldSyncDebugger>, sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh> > const, sierra::nalu::nalu_ngp::impl::NodeFieldOp<stk::mesh::Device\
Mesh, stk::mesh::DeviceField<double, stk::mesh::EmptyNgpFieldSyncDebugger>, sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh> > const, sierra::nalu::nalu_ngp::impl::NodeFieldOp<stk::mesh::DeviceMesh, stk::mesh::DeviceField<double, stk::mesh::EmptyNgpFieldSyncDebugger>, sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh> > const> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, sierra::nalu::nalu_ngp::MeshInfo<stk::\
mesh::DeviceMesh, sierra::nalu::nalu_ngp::FieldManager> const&, sierra::nalu::ElemDataRequests const&, sierra::nalu::ElemDataRequests const&, stk::mesh::Selector const&, __nv_hdl_wrapper_t<false, false, __nv_dl_tag<void (sierra::nalu::WallFuncGeometryAlg<sierra::nalu::AlgTraitsFaceElem<sierra::nalu::AlgTraitsEdge_2D, sierra::nalu::AlgTraitsQuad4_2D> >::*)(), &sierra::nalu::WallFuncGeometryAlg<sierra::nalu::AlgTraitsFaceElem<sierra::nalu::AlgTraitsEdge_2D, sierra::nalu::AlgTraitsQu\
ad4_2D> >::execute, 1u>, void (sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh>&), unsigned int const, unsigned int const, sierra::nalu::MasterElement*, sierra::nalu::MasterElement*, bool, double, sierra::nalu::nalu_ngp::impl::ElemFieldOp<stk::mesh::DeviceMesh, stk::mesh::DeviceField<double, stk::mesh::EmptyNgpFieldSyncDebugger>, sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh> > const, sierra::nalu::nalu_ngp::impl::NodeFieldOp<stk::mesh::DeviceMesh, \
stk::mesh::DeviceField<double, stk::mesh::EmptyNgpFieldSyncDebugger>, sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh> > const, sierra::nalu::nalu_ngp::impl::NodeFieldOp<stk::mesh::DeviceMesh, stk::mesh::DeviceField<double, stk::mesh::EmptyNgpFieldSyncDebugger>, sierra::nalu::nalu_ngp::FaceElemSimdData<stk::mesh::DeviceMesh> > const>) (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==    by 0x1113659F: sierra::nalu::WallFuncGeometryAlg<sierra::nalu::AlgTraitsFaceElem<sierra::nalu::AlgTraitsEdge_2D, sierra::nalu::AlgTraitsQuad4_2D> >::execute() (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==    by 0x112C63B3: sierra::nalu::NgpAlgDriver::execute() (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==    by 0x105C7BD7: sierra::nalu::Realm::initialize_prolog() (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==    by 0x105E0913: sierra::nalu::Realms::initialize_prolog() (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==    by 0x10607EF3: sierra::nalu::Simulation::initialize() (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==    by 0x101018AB: main (in /gpfs/alpine/cfd116/scratch/mullowne/nalu-wind/build_gpu_tr_develop_release_hypre_master/naluX)
==3068855==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==3068855==
==3068855==
==3068855== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==3068855==    at 0xFF18E20: raise (in /usr/lib64/power9/libpthread-2.28.so)
==3068855==    by 0x571B617B: show_stackframe (in /autofs/nccs-svm1_sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-6depextb6p6ulrvmehgtbskbmcsyhtdi/lib/libopen-pal.so.3.1.1)
==3068855==    by 0x5807F133: ??? (in /autofs/nccs-svm1_sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-8.3.1/valgrind-3.17.0-tww4ja3hdejwwedeagwr4jnwfwkdljua/libexec/valgrind/memcheck-ppc64le-linux)
airfoilRANSEdgeNGPHypre.rst doesn't segfault with RELWITHDEBINFO, it does with a RELEASE build. The valgrind output is informative. In the appropriate directory, I run:

**Comment by PaulMullowney** [comment]
[airfoilRANSEdgeNGPTrilinos.rst.log](https://github.com/Exawind/nalu-wind/files/7906840/airfoilRANSEdgeNGPTrilinos.rst.log)
[trilinos_vg.txt](https://github.com/Exawind/nalu-wind/files/7906849/trilinos_vg.txt)

**Comment by alanw0** [comment]
We'll see if we can get someone else to reproduce this.

**Comment by PaulMullowney** [comment]
I'm trying to compile with a single thread so we don't see the mangled output. This may take a while.

**Comment by alanw0** [comment]
I'm not sure what mangled output you're referring to. The mangling of the symbol names?

**Comment by PaulMullowney** [comment]
Sorry wrong issue.

**Comment by jrood-nrel** [comment]
Are they the same segfaults as on the Nalu-Wind tests on Eagle?

**Comment by alanw0** [comment]
@jrood-nrel yes they do look the same. And the weird thing about the eagle ones, a few days ago I didn't see any segfaults, only diffs, but now yesterday there are failures that have this segfault. At least if they were there before I didn't see them.
Anyway, Timothy and I determined that the segfault is happening in the nodes_per_entity function when nalu tries to get num-nodes from the master-element on device. We referred it to James. James @overfelt do you have any idea what might be happening here?

**Comment by jrood-nrel** [comment]
That is because the GPU tests were switched from GCC 8.4.0 to 9.3.0 and that's when this started happening. I think they work with GCC 8.

**Comment by alanw0** [comment]
ok that makes sense. In that case, we might have better luck debugging this on our own cuda machine here, depending on what James has found.

**Comment by overfelt** [comment]
I have only just managed to replicate the problem on summit and still do not get anything out of valgrind that is helpful.

**Comment by PaulMullowney** [comment]
This may be related.
https://github.com/Exawind/nalu-wind/issues/816

**Comment by overfelt** [comment]
Yes, that looks like the same error.  The conversation seemed so promising until the last entry which says the problem just went away for no obvious reason so the ticket was closed.  Seems to be back again.

**Comment by PaulMullowney** [comment]
Shreyas pointed out a PR for a similar issue.

https://github.com/Exawind/nalu-wind/pull/384

**Comment by PaulMullowney** [comment]
The fix suggested in #816 seems to enable nalu to run those reg tests on summit when building in Release mode

**Comment by jrood-nrel** [comment]
Sounds like a stack issue if bypassing a function call fixes it. Did you try setting the cuda stack size?

**Comment by PaulMullowney** [comment]
I haven't tried that yet. I'll do that next.

**Comment by alanw0** [comment]
There are some calls to cudaDeviceSetLimit in Simulation.C.
I know Timothy and Dong Hun had encountered those the other day when trying to build for AMD/HIP.

**Comment by PaulMullowney** [comment]
Perhaps this should a parameter that can be changed in the input file? I could add that so its configurable.

**Comment by alanw0** [comment]
> Perhaps this should a parameter that can be changed in the input file? I could add that so its configurable.

Yes I think that's a good idea. I wonder if there's an equivalent call for AMD/HIP.

**Comment by overfelt** [comment]
Unfortunately there is little in the way of suggestions on how to set the stack size or when to change it.  If your run dies without any kind of meaningful error then maybe increasing this number will fix it.  Since it is based on how much memory is needed for the most memory intensive Cuda kernel launch, it is hoped that a good suite of regression tests that stress the kernels would catch it.

**Comment by PaulMullowney** [comment]
I think we should be testing nrel5mw mesh nightly. The big version with 23million mesh nodes requires 12 GPUs. I bet we could do the smaller version (7 million mesh nodes) on about 4 GPUs.

**Comment by overfelt** [comment]
No, I think exhausting  the stack has more to do with compile time static memory in kernels.  In the SPARC project we see this because there are kernels templated on sizes that can grow as we add more functionality.  For Nalu it might be a kernel that is templated on something like the number of integration points that then has lots of fixed arrays of that size.  Memory allocated by Kokkos to hold dynamic arrays can check if the allocated pointer is null which I'm sure Kokkoos does. I have not googled this in a while, maybe these days there is a tool that can spit out the maximum amount of stack size used at the end of a run so at least we would know if we are getting close to the max.

**Comment by jhux2** [comment]
Someone on the Kokkos slack channel may be able to give guidance on this.

**Comment by PaulMullowney** [comment]
I've tried 4096 ... 65536 in powers of 2. None succeed.

**Comment by PaulMullowney** [comment]
I see this when running the hybrid solver on the sphere test using racecheck compute-sanitizer tool.

Kokkos is the issue.

========= WARNING: (Warp Level Programming) Potential RAW hazard detected at __shared__ 0x3ba in block (0,0,0) :
=========     Write Thread (0,238,0) at 0x11d0 in /gpfs/alpine/cfd116/proj-shared/mullowne/spack-manager/spack/opt/spack/linux-rhel8-power9le/gcc-9.3.0/trilinos-develop-gr25dapfwj7q3uar43nudpjbd2odxit6/include/Cuda/Kokkos_Cuda_ReduceScan.hpp:544:__nv_static_60__47_tmpxft_00159400_00000000_7_ScratchViews_cpp1_ii_8383be68__ZN6Kokkos4Impl33cuda_parallel_launch_local_memoryINS0_14ParallelReduceINS0_18CudaFunctorAdapterIZN6sierra4nalu16nodes_per_entityINS5_19ElemDataRequestsGPUEEEiRKT_NS5_6METypeEEUliRiE_NS_11RangePolicyIJNS_4CudaEEEEivEESG_NS_11InvalidTypeESF_EEEEvS8_
=========     Read Thread (0,236,0) at 0x1270 in /gpfs/alpine/cfd116/proj-shared/mullowne/spack-manager/spack/opt/spack/linux-rhel8-power9le/gcc-9.3.0/trilinos-develop-gr25dapfwj7q3uar43nudpjbd2odxit6/include/Cuda/Kokkos_Cuda_ReduceScan.hpp:544:__nv_static_60__47_tmpxft_00159400_00000000_7_ScratchViews_cpp1_ii_8383be68__ZN6Kokkos4Impl33cuda_parallel_launch_local_memoryINS0_14ParallelReduceINS0_18CudaFunctorAdapterIZN6sierra4nalu16nodes_per_entityINS5_19ElemDataRequestsGPUEEEiRKT_NS5_6METypeEEUliRiE_NS_11RangePolicyIJNS_4CudaEEEEivEESG_NS_11InvalidTypeESF_EEEEvS8_
=========     Current Value : 0
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame: [0x20005118080c]
=========                in /lib64/libcuda.so.1

**Comment by PaulMullowney** [comment]
Here' some more. 

========= WARNING: Race reported between Read access at 0x19c0 in /gpfs/alpine/cfd116/proj-shared/mullowne/spack-manager/spack/opt/spack/linux-rhel8-power9le/gcc-9.3.0/trilinos-develop-gr25dapfwj7q3uar43nudpjbd2odxit6/include/Cuda/Kokkos_Cuda_ReduceScan.hpp:548:__nv_static_60__47_tmpxft_00159400_00000000_7_ScratchViews_cpp1_ii_8383be68__ZN6Kokkos4Impl33cuda_parallel_launch_local_memoryINS0_14ParallelReduceINS0_18CudaFunctorAdapterIZN6sierra4nalu16nodes_per_entityINS5_19ElemDataRequ\
estsGPUEEEiRKT_NS5_6METypeEEUliRiE_NS_11RangePolicyIJNS_4CudaEEEEivEESG_NS_11InvalidTypeESF_EEEEvS8_
=========     and Write access at 0x1920 in /gpfs/alpine/cfd116/proj-shared/mullowne/spack-manager/spack/opt/spack/linux-rhel8-power9le/gcc-9.3.0/trilinos-develop-gr25dapfwj7q3uar43nudpjbd2odxit6/include/Cuda/Kokkos_Cuda_ReduceScan.hpp:544:__nv_static_60__47_tmpxft_00159400_00000000_7_ScratchViews_cpp1_ii_8383be68__ZN6Kokkos4Impl33cuda_parallel_launch_local_memoryINS0_14ParallelReduceINS0_18CudaFunctorAdapterIZN6sierra4nalu16nodes_per_entityINS5_19ElemDataRequestsGPUEEEiRKT_NS5_6ME\
TypeEEUliRiE_NS_11RangePolicyIJNS_4CudaEEEEivEESG_NS_11InvalidTypeESF_EEEEvS8_ [4 hazards]
=========     and Write access at 0x19e0 in /gpfs/alpine/cfd116/proj-shared/mullowne/spack-manager/spack/opt/spack/linux-rhel8-power9le/gcc-9.3.0/trilinos-develop-gr25dapfwj7q3uar43nudpjbd2odxit6/include/Cuda/Kokkos_Cuda_ReduceScan.hpp:548:__nv_static_60__47_tmpxft_00159400_00000000_7_ScratchViews_cpp1_ii_8383be68__ZN6Kokkos4Impl33cuda_parallel_launch_local_memoryINS0_14ParallelReduceINS0_18CudaFunctorAdapterIZN6sierra4nalu16nodes_per_entityINS5_19ElemDataRequestsGPUEEEiRKT_NS5_6ME\
TypeEEUliRiE_NS_11RangePolicyIJNS_4CudaEEEEivEESG_NS_11InvalidTypeESF_EEEEvS8_ [4 hazards]

**Comment by PaulMullowney** [comment]
I don't know what this means either. However, I'll point out that it's pointing at nodes_per_entity and it's happening despite the "fix" that I put in order to get the regression tests to run.

**Comment by PaulMullowney** [comment]
I think I've got a workaround for this problem. I've shown that it stops the segfaulting for at least 1 test. I want to try to remove the compute-sanitizer errors by generalizing my solution.

**Comment by PaulMullowney** [comment]
My workaround for this seems to be doing ok. All Hypre GPU regression tests run on 4 GPUs on summit when built in release. The race check warnings that I fixed in nodes_per_entity have now moved into num_integration_points method.

I've tried 2 Trilinos tests. They run on a single GPU however they crash in MPI code on multiple GPUs. I'm not that well versed in Trilinos solver class to fix this. Could someone volunteer to take this on?
